spopt/tests/test_p_dispersion.py

```diff
@@ -0,0 +1,222 @@
+from pyproj import crs
+from spopt.locate.base import FacilityModelBuilder, LocateSolver, T_FacModel
+import numpy
+import geopandas
+import pandas
+import pulp
+import spaghetti
+from shapely.geometry import Point, Polygon
+
+from spopt.locate import PDispersion
+from spopt.locate.util import simulated_geo_points
+import unittest
+import os
+import pickle
+import platform
+
+operating_system = platform.platform()[:7].lower()
+if operating_system == "windows":
+    WINDOWS = True
+else:
+    WINDOWS = False
+
+
+class TestSyntheticLocate(unittest.TestCase):
+    def setUp(self) -> None:
+        self.dirpath = os.path.join(os.path.dirname(__file__), "./data/")
+
+        lattice = spaghetti.regular_lattice((0, 0, 10, 10), 9, exterior=True)
+        ntw = spaghetti.Network(in_data=lattice)
+        gdf = spaghetti.element_as_gdf(ntw, arcs=True)
+        street = geopandas.GeoDataFrame(
+            geopandas.GeoSeries(gdf["geometry"].buffer(0.2).unary_union),
+            crs=gdf.crs,
+            columns=["geometry"],
+        )
+
+        facility_count = 5
+
+        self.facility_points = simulated_geo_points(
+            street, needed=facility_count, seed=6
+        )
+
+        ntw = spaghetti.Network(in_data=lattice)
+
+        ntw.snapobservations(self.facility_points, "facilities", attribute=True)
+
+        self.facilities_snapped = spaghetti.element_as_gdf(
+            ntw, pp_name="facilities", snapped=True
+        )
+
+        self.cost_matrix = ntw.allneighbordistances(
+            sourcepattern=ntw.pointpatterns["facilities"],
+            destpattern=ntw.pointpatterns["facilities"],
+        )
+
+
+    def test_p_dispersion_from_cost_matrix(self):
+        pdispersion = PDispersion.from_cost_matrix(self.cost_matrix, p_fac=2)
+        result = pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+        self.assertIsInstance(result, PDispersion)
+
+    #not sure if i'll need this... or how could the test be retrofitted to work well with this class?
+    '''    def test_p_dispersion_facility_client_array_from_cost_matrix(self):
```

Comment:
Are these commented out tests repeats from the above ones?

Reply:
Basically. One set is for solving from a cost matrix and the other is for solving from a geodataframe. I left them in the file and commented out because I wanted to be sure we didn't need anything similar to them, but it's pretty clear now they have no relevance so I'll delete.

Reply:
Hmmm. Maybe don't delete just yet. We do want to test all functionality.

Reply:
I think we should discuss how to make these tests more general. I mean create a class to test the `from_cost_matrix` and `from_geodataframe` methods to decrease the test code's length. But not for this PR :)